allensdk/brain_observatory/behavior/behavior_project_lims_api.py

```diff
@@ -32,54 +35,87 @@ def __init__(self, postgres_engine, app_engine):
+                    exactly one column in the response, return the values as a
+                    list.
         app_engine :
             used for making queries agains the lims web application. Must
             implement:
                 stream : takes a url as a string. Returns an iterable yielding
                 the response body as bytes.
         """
-        self.postgres_engine = postgres_engine
+        self.lims_engine = lims_engine
+        self.mtrain_engine = mtrain_engine
         self.app_engine = app_engine
         self.logger = logging.getLogger("BehaviorProjectLimsApi")
 
     @classmethod
     def default(
         cls,
-        pg_kwargs: Optional[Dict[str, Any]] = None,
+        lims_credentials: Optional[DbCredentials] = None,
+        mtrain_credentials: Optional[DbCredentials] = None,
         app_kwargs: Optional[Dict[str, Any]] = None) -> \
             "BehaviorProjectLimsApi":
         """Construct a BehaviorProjectLimsApi instance with default
         postgres and app engines.
 
-        :param pg_kwargs: dict of keyword arguments to pass to the
-        PostgresQueryMixin class instance. Valid arguments include:
-            "dbname", "user", "host", "password", "port". Will use
-            defaults in PostGresQueryMixin.__init__ if unspecified.
-        :type pg_kwargs: dict
-        :param app_kwargs: dict of keyword arguments to pass to the
-            HTTPEngine class instance. Valid arguments include:
-            "scheme", "host". Will default to scheme=http, host=lims2
-            if left unspecified.
-        :type app_kwargs: dict
-        :rtype: BehaviorProjectLimsApi
+        Parameters
+        ----------
+        lims_credentials: Optional[DbCredentials]
+            Credentials to pass to the postgres connector to the lims database.
+            If left unspecified, will check environment variables for the
+            appropriate values.
+        mtrain_credentials: Optional[DbCredentials]
+            Credentials to pass to the postgres connector to the mtrain
+            database. If left unspecified, will check environment variables
+            for the appropriate values.
+        app_kwargs: Dict
+            Dict of arguments to pass to the app engine. Currently unused.
+
+        Returns
+        -------
+        BehaviorProjectLimsApi
         """
-        _pg_kwargs = pg_kwargs or dict()
 
         _app_kwargs = {"scheme": "http", "host": "lims2"}
         if app_kwargs:
             _app_kwargs.update(app_kwargs)
+        if lims_credentials:
+            lims_engine = PostgresQueryMixin(
+                dbname=lims_credentials.dbname, user=lims_credentials.user,
+                host=lims_credentials.host, password=lims_credentials.password,
+                port=lims_credentials.port)
+        else:
+            # Currying is equivalent to decorator syntactic sugar
+            lims_engine = (credential_injector(LIMS_DB_CREDENTIAL_MAP)
```

Comment:
I'm not seeing this used as a decorator anywhere. Is that right? Maybe define the wrapped function outside the decorator so it can be separately imported to cut down on syntactical noise?

Reply:
I'm not using the decorator syntactic sugar (since I'm not decorating function definitions), but I do use the equivalent functionality via currying. What do you mean by syntactical noise?

Reply:
Do you mean defining a separate function, say, "create_lims_connection" that is decorated with the credential injector, and called by this class?

Reply:
I just mean that you've defined credential_injector as a decorator to gain some syntactic benefit, but all of your use cases just want a second-order function, so c_i being defined as a decorator only exacts a syntactic cost:
```
credential_injector(LIMS_DB_CREDENTIAL_MAP)(PostgresQueryMixin)()
vs.
something_similar(LIMS_DB_CREDENTIAL_MAP, PostgresQueryMixin)
```
Not crucial, of course.

Reply:
Ah, I see your point. My original design had it decorating the PostgresQueryMixin `__init__` function, but since our use-case requires multiple valid sets of credentials (e.g. lims and mtrain) I wasn't able to do it that way. I kept the general design though because I thought having a generic decorator might be useful in the future, but that may be thinking too far ahead.

Do you think it's confusing enough to warrant a rewrite?